docs/development/extensions-contrib/k8s-jobs.md

```diff
@@ -0,0 +1,125 @@
+---
+id: k8s-jobs
+title: "MM-less Druid in K8s"
+---
+
+<!--
+  ~ Licensed to the Apache Software Foundation (ASF) under one
+  ~ or more contributor license agreements.  See the NOTICE file
+  ~ distributed with this work for additional information
+  ~ regarding copyright ownership.  The ASF licenses this file
+  ~ to you under the Apache License, Version 2.0 (the
+  ~ "License"); you may not use this file except in compliance
+  ~ with the License.  You may obtain a copy of the License at
+  ~
+  ~   http://www.apache.org/licenses/LICENSE-2.0
+  ~
+  ~ Unless required by applicable law or agreed to in writing,
+  ~ software distributed under the License is distributed on an
+  ~ "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+  ~ KIND, either express or implied.  See the License for the
+  ~ specific language governing permissions and limitations
+  ~ under the License.
+  -->
+
+Consider this an [EXPERIMENTAL](../experimental.md) feature mostly because it has not been tested yet on a wide variety of long-running Druid clusters.
+
+Apache Druid Extension to enable using Kubernetes for launching and managing tasks instead of the Middle Managers.  This extension allows you to launch tasks as K8s jobs removing the need for your middle manager.  
+
+## How it works
+
+It takes the podSpec of your `Overlord` pod and creates a kubernetes job from this podSpec.  Thus if you have sidecars such as splunk, hubble, istio it can optionally launch a task as a k8s job.  All jobs are natively restorable, they are decopled from the druid deployment, thus restarting pods or doing upgrades has no affect on tasks in flight.  They will continue to run and when the overlord comes back up it will start tracking them again.  
```

Comment:
There seems to be one pod spec for a cluster. Should I then ensure that the pod is large enough for my largest task? Would a future enhancement be the ability to override the pod spec per-task so that light-weight tasks (the MSQ controller) takes less resources than a heavy-weight task (an MSQ worker)?

Reply:
So the reason for the podSpec is that I didn't want to take a brand new docker image and create a job spec for it.  I don't know your cluster config.  Suppose you have secrets mounted, volumes setup up your way, env variables, certs, annotations for things like istio, etc... The parent pod spec logic is to  ensure you have those things for your peon task and you don't have to worry about it.  But, I do take that pod spec and massage it.  I do things like change the resources required.  For CPU we always give the task a single core (just like what druid was always doing).  For memory we take the jvm opts you pass and configure the container resources from that.  I believe its something like (Xmx + dbb)*1.2.  So while the cpu resources are fixed at a core, the memory is deduced from your jvm opts.  I want to make sure I answered your question correctly here, please let me know things don't make sense.

Reply:
> Perhaps provide just a bit more background for newbies. Where does the pod spec come from? Something like:
> 
> > With this extension, your Overload tasks run in a Kubernetes pod. You define that pod using a pod spec. The pod spec names a Docker image with Druid installed, identifies K8s settings, provides Druid configuration, passes in secrets, etc. You create the pod spec outside of Druid. We suggest you test the spec directly in K8s before using it with Druid.
> 
> I'm sure some (all?) of the above is wrong: it just identifies what I, as a newbie to this extension, would need to know. Focus in particular on anything Druid-specific I need to provide in the pod spec.
> 
> Also, where do I put the pod spec? In K8s somewhere? On the local disk of each Druid node?

Maybe I wasn't clear, you don't have to do anything here.  The podSpec comes from the overlord pod, the pod that launches the task.  The `K8sTaskAdapter` takes the overlord spec and modifies it to run a peon k8s job.  The user has to do nothing here.

Reply:
> I'm sure some (all?) of the above is wrong: it just identifies what I, as a newbie to this extension, would need to know. Focus in particular on anything Druid-specific I need to provide in the pod spec.
> 
> Also, where do I put the pod spec? In K8s somewhere? On the local disk of each Druid node?

Should i just exclude the podSpec part?  Users have to do nothing here, I am just explaining how it works at a high-level.

Reply:
it makes sense to get the podspec from the overlord. But I suppose that I can override some elements in the spec if I want as a user?

Reply:
yes definitely.  Right now I give you the ability to override annotations and labels.  In a later PR we can make this more configurable for users.  I have used this patch for a few druid clusters and did not need to override anything else, but that is just my experience.  I think going forward we make  this more configurable in the future.  The goal of this patch was to make it as easy as possible for users to switch from using MM's to k8s jobs.  Currently its add 4 required configurations for the overlord and you are good to go.

Reply:
For sure, I don't think we need those features right away.